storage/engine.go

```diff
@@ -18,17 +18,103 @@
 
 package storage
 
-import "github.com/nuts-foundation/nuts-node/core"
+import (
+	"context"
+	"errors"
+	"github.com/nuts-foundation/go-stoabs"
+	"github.com/nuts-foundation/go-stoabs/bbolt"
+	"github.com/nuts-foundation/nuts-node/core"
+	"github.com/nuts-foundation/nuts-node/storage/log"
+	"path"
+	"strings"
+	"time"
+)
+
+const storeShutdownTimeout = 5 * time.Second
 
 // New creates a new instance of the storage engine.
-func New() core.Engine {
-	return &engine{}
+func New() Engine {
+	return &engine{stores: map[string]stoabs.Store{}}
 }
 
 type engine struct {
+	datadir string
+	stores  map[string]stoabs.Store
 }
 
-// Name returns the name of the storage engine.
+// Name returns the name of the engine.
 func (e engine) Name() string {
 	return "Storage"
 }
+
+func (e engine) Start() error {
+	return nil
+}
+
+func (e engine) Shutdown() error {
+	shutdown := func(store stoabs.Store) error {
+		// Refactored to separate function, otherwise defer would be in for loop which leaks resources.
+		ctx, cancel := context.WithTimeout(context.Background(), storeShutdownTimeout)
+		defer cancel()
+		return store.Close(ctx)
+	}
+
+	failures := false
+	for storeName, store := range e.stores {
+		err := shutdown(store)
+		if err != nil {
+			log.Logger().Errorf("Failed to close store '%s': %s", storeName, err)
+			failures = true
+		}
+	}
+	if failures {
+		return errors.New("one or more stores failed to close")
+	}
+	return nil
+}
+
+func (e *engine) Configure(config core.ServerConfig) error {
+	e.datadir = config.Datadir
+	return nil
+}
+
+func (e *engine) GetProvider(moduleName string) Provider {
+	return &provider{
+		moduleName: strings.ToLower(moduleName),
+		engine:     e,
+	}
+}
+
+type provider struct {
+	moduleName string
+	engine     *engine
+}
+
+func (p *provider) GetKVStore(name string) (stoabs.KVStore, error) {
+	store, err := p.getStore(p.moduleName, name, func(moduleName string, name string) (stoabs.Store, error) {
+		return bbolt.CreateBBoltStore(path.Join(p.engine.datadir, moduleName, name+".db"))
+	})
+	if store == nil {
+		return nil, err
+	}
+	return store.(stoabs.KVStore), err
+}
+
+func (p *provider) getStore(moduleName string, name string, creator func(namespace string, name string) (stoabs.Store, error)) (stoabs.Store, error) {
```

Comment:
Probably not an issue when starting the node:
not thread safe, requesting the same store in parallel would cause one of the callers to lock

Reply:
I considered this but came to the same conclusion. But indeed, an easy fix 👍